### PR TITLE
fix: Correct handling of filter and selectionFlags

### PR DESCRIPTION
**Pull request by tylerbutler**
## Description

The `--dir` flag in the `exec` command is not working, and the root cause is that the code expected the flag to be called `directory` but it was called `dir`. The reason this wasn't caught was because of sloppy typing. I had used an `any` to parse the raw flags into filter and selection objects, which introduced this error.

## Fix

The flag types are somewhat complex, so it seemed simpler to make simpler "mirror types" that structurally match the oclif flag types, but are typed more simply. This allows me to add some type safety to the parse functions.

## Testing

I added several new tests that exercise the selection and filtering. Most tests work against the helper filtering functions. However, there are also tests that run through a new test-only command to validate that converting oclif flags into filter/selection objects is done correctly.

### CODE REVIEW

File: build-tools/packages/build-cli/src/filter.ts [hunk 1]
```diff
@@ -51,12 +52,13 @@ export interface PackageFilterOptions {
 	private: boolean | undefined;
 }
 
+// {dir: '.', verbose: false, quiet: false, timer: false, concurrency: 25, …}
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/filter.ts:55]:
```suggestion
```

File: build-tools/packages/build-cli/src/flags.ts [hunk 2]
```diff
@@ -253,6 +253,21 @@ export const selectionFlags = {
...
 };
 
+/**
+ * This interface is used for type enforcement of selection flags. The default oclif typing is complex and difficult to
+ * simplify, so this type just mirrors the object above with simpler typing. This should ONLY be used when processing
+ * raw flags.
+ *
+ * @internal
+ */
+export interface selectionFlags {
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/src/flags.ts:263]: I suspect all these fields can be made readonly.

Comment (CraigMacomber) [reply build-tools/packages/build-cli/src/flags.ts:263]: also the arrays inside can also prabably be readonly.

Comment (tylerbutler) [reply build-tools/packages/build-cli/src/flags.ts:263]: Good point. I'll update these.

File: build-tools/packages/build-cli/src/commands/test-only-filter.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { Package } from "@fluidframework/build-tools";
+import { strict as assert } from "node:assert";
+// import path from "node:path";
+
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/test-only-filter.ts:8]:
```suggestion

```

File: build-tools/packages/build-cli/src/index.ts [hunk 4]
```diff
@@ -4,4 +4,4 @@
  */
 
 export { run } from "@oclif/core";
-export type { ReleaseGroup, ReleasePackage } from "./releaseGroups";
+export type { knownReleaseGroups, ReleaseGroup, ReleasePackage } from "./releaseGroups";
```
Comment (alexvy86) [inline build-tools/packages/build-cli/src/index.ts:7]: Does it need to be exported?

File: build-tools/packages/build-cli/src/BasePackageCommand.ts [hunk 5]
```diff
@@ -34,8 +35,18 @@ export abstract class PackageCommand<
...
 	};
 
-	private filterOptions: PackageFilterOptions | undefined;
-	private selectionOptions: PackageSelectionCriteria | undefined;
+	protected filterOptions: PackageFilterOptions | undefined;
+	protected selectionOptions: PackageSelectionCriteria | undefined;
+
+	/**
+	 * An array of packages selected based on the selection criteria.
+	 */
```
Comment (alexvy86) [inline build-tools/packages/build-cli/src/BasePackageCommand.ts:43]: Thoughts?
```suggestion
	/**
	 * An array of packages selected based on the selection criteria.
	 *
	 * @remarks
	 * These are not necessarily the packages that get acted on.
	 * Filter criteria are first applied to them to determine that.
	 */
```

**Review by alexvy86 — APPROVED** [review]
I only skimmed the tests, but overall lgtm. A few nits below.

File: build-tools/packages/build-cli/src/index.ts (hunk 4 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/index.ts:7]: Alas, it does. I don't remember why I export the ReleaseGroup type, but I do, so I have to export this as well. I don't think any of it needs to exported, but that's a breaking change (though practically it's probably not a big deal) so I skipped it.

File: build-tools/packages/build-cli/src/BasePackageCommand.ts (hunk 5 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/BasePackageCommand.ts:43]: Good idea. I'll add these details, just want to tweak the wording.

**Pull request merged**